Finish early mono sign when skipping factor

micbakos-rdx opened the pull request:
Updated mono sign, to return no request when there are no remaining per transaction requests. When a user skips a factor source, the signature collector will iterate to the next factor source. There if there are no more remaining still to be signed as valid transactions, the iterator continues. This ends up ending the collector's signing process, with an unsuccessful outcome.

* Updated the tests to incorporate skipping of a factor source instead of just failing one.
* Added some missing kotlin extensions needed in host to sign, with a mnemonic, multiple per transaction signables at once and also an extension to derive factor source ids. Both were required for the integration with the host

matiasbzurovski commented on crates/app/signing/src/collector/signatures_collector.rs:
```diff
@@ -253,30 +253,33 @@ impl<S: Signable> SignaturesCollector<S> {
 
-            if self.continuation() == FinishEarly {
-                break;
+                if self.continuation() == FinishEarly {
+                    break;
```
shouldn't we also check this when there is no request for the given `factor_source_id`. This, when there are no transactions to sign for the given factor source

CyonAlexRDX commented on crates/app/signing-traits/src/testing/simulated_user.rs:
```diff
@@ -99,31 +131,37 @@ impl SimulatedUserMode {
+    }
+
+    pub fn prudent_with_skips(simulated_skips: SimulatedSkips) -> Self {
+        Self::new(SimulatedUserMode::Prudent, None, simulated_skips)
```
`Prudent` means "never skips". So this looks wrong. Better let this be ctor of `Lazy`. And update definition of Lazy:

```diff
-    /// Emulation of a "lazy" user, that skips signing with as many factor
-    /// sources as possible.
+    /// Emulation of a "lazy" user, that skips signing with as some factor
+    /// sources, depending on `Laziness`.
    Lazy(Laziness),
```

micbakos-rdx replied:
Yes, but `Laziness` has no option to sign all the rest of the factor sources except the ones defined in `simulated_skips`. This is what I wanted to achieve. 

Lazyness describes a user that either skips all or signs the minimum.

Sajjon replied:
Yes so we update Laziness enum with a third variant..?

SkipsSpecific(IndexSet<_>)

Sajjon commented on crates/system/os/signing/src/sargon_os_signing.rs:
```diff
@@ -418,6 +504,7 @@ mod test {
 
     enum SigningFailure {
         NeglectedFactorSources(Vec<FactorSourceIDFromHash>),
+        SkippingFactorSources(Vec<FactorSourceIDFromHash>),
```
We used to define "Skip(ped)" a subset of "Neglected" - so is this needed?

Sajjon replied:
You can probably use Neglected variants

micbakos-rdx replied:
I need to rename `NeglectedFactorSources` to `FailingFactorSources`

micbakos-rdx commented:
@matiasbzurovski @CyonAlexRDX 

Latest changes:

1. Added to user mode a user that skips specific factors.
2. Introduced a new error for when the user fails to sign since too many factors neglected.
3. Added `--locked` to cargo ndk.
